sycl/source/detail/queue_impl.hpp

```diff
@@ -240,6 +240,13 @@ class queue_impl {
     RT::PiContext Context = MContext->getHandleRef();
     RT::PiDevice Device = MDevice->getHandleRef();
     const detail::plugin &Plugin = getPlugin();
+    if (!(Plugin == MDevice->getPlugin())) {
+      throw cl::sycl::invalid_object_error(
```

Comment:
can you explain how is it possible that the MDevice & MContext are from different BE/plugins?

Reply:
I accidentally found this while writing some sample code, but also found this was the reason behind a LIT test segfault.
The following SYCL queue constructor (in https://github.com/intel/llvm/blob/sycl/sycl/include/CL/sycl/queue.hpp#L113)

```cpp
  queue(const context &SyclContext, const device &SyclDevice,
        const property_list &PropList = {});
```

creates a queue using an existing SYCL context and device.
If the SYCL context comes from a different backend than the device, the PI plugin that is called is different from the backend from the passed device and the PI call segfaults.

This checks provides a better error message.

Reply:
On the line 75 in this file we have a check that a context passed in bound to a device, I guess this check should prevent the situation you are fixing here. But we call createQueue before doing this check, probably we should move the call after the check(line 75)?

Reply:
This error message is more specific, but maybe that is not that important? Up to you. I think there are other places where checking for backends is needed?

Reply:
+1 to move the existing check on line 75 before the call to `createQueue`.
And, I agree, we should have similar checks in all interfaces that take multiple SYCL objects (that their plugins are the same)

Reply:
I've reordered the lines, and changed the check to an assert